src/cryptography/x509/name.py

```diff
@@ -196,14 +196,32 @@ def __repr__(self) -> str:
 
 
 class Name(object):
-    def __init__(self, attributes):
+    @typing.overload
+    def __init__(self, attributes: typing.Iterable[NameAttribute]) -> None:
+        ...
+
+    @typing.overload
+    def __init__(
+        self, attributes: typing.Iterable[RelativeDistinguishedName]
+    ) -> None:
+        ...
+
+    def __init__(
+        self,
+        attributes: typing.Iterable[
```

Comment:
Would writing this as `typing.Union[typing.Iterable[NameAttribute], typing.Iterable[RelativeDistinguishedName]` allow us to avoid the need for the overloads?

Reply:
I did try this, and you get:

```
$ mypy src/cryptography/ vectors/cryptography_vectors/ tests/
src/cryptography/x509/name.py:206: error: Incompatible types in assignment (expression has type "List[object]", variable has type "Union[Iterable[NameAttribute], Iterable[RelativeDistinguishedName]]")
```

When I typehint the attributes cast, I get:

```
$ mypy src/cryptography/ vectors/cryptography_vectors/ tests/
src/cryptography/x509/name.py:206: error: Name 'attributes' already defined on line 199  [no-redef]
```

Also when using a different name in the assignment I again get "Incompatible types in assingment". So... I'm open to any way to improve that. 

For the record, here is the full change I tried:

```diff
diff --git a/src/cryptography/x509/name.py b/src/cryptography/x509/name.py
index 0bfc2d94..f5be2b06 100644
--- a/src/cryptography/x509/name.py
+++ b/src/cryptography/x509/name.py
@@ -196,31 +196,25 @@ class RelativeDistinguishedName(object):
 
 
 class Name(object):
-    @typing.overload
-    def __init__(self, attributes: typing.Iterable[NameAttribute]) -> None:
-        ...
-
-    @typing.overload
-    def __init__(
-        self, attributes: typing.Iterable[RelativeDistinguishedName]
-    ) -> None:
-        ...
-
     def __init__(
         self,
-        attributes: typing.Iterable[
-            typing.Union[NameAttribute, RelativeDistinguishedName]
+        attributes: typing.Union[
+            typing.Iterable[NameAttribute],
+            typing.Iterable[RelativeDistinguishedName],
         ],
     ) -> None:
-        attributes = list(attributes)
-        if all(isinstance(x, NameAttribute) for x in attributes):
+        attrs: typing.Union[
+            typing.List[NameAttribute],
+            typing.List[RelativeDistinguishedName],
+        ] = list(attributes)
+        if all(isinstance(x, NameAttribute) for x in attrs):
             self._attributes = [
                 RelativeDistinguishedName([typing.cast(NameAttribute, x)])
-                for x in attributes
+                for x in attrs
             ]
-        elif all(isinstance(x, RelativeDistinguishedName) for x in attributes):
+        elif all(isinstance(x, RelativeDistinguishedName) for x in attrs):
             self._attributes = typing.cast(
-                typing.List[RelativeDistinguishedName], attributes
+                typing.List[RelativeDistinguishedName], attrs
             )
         else:
             raise TypeError(
```

Reply:
What a mess, thanks.

Reply:
Noticed before you really don't like these overloads. They look quite okay to me.